core/application/src/state/executor/epoch_change.rs

```diff
@@ -73,107 +76,635 @@ impl<B: Backend> StateExecutor<B> {
         self.committee_info
             .set(current_epoch, current_committee.clone());
 
-        // If more than 2/3rds of the committee have signaled, start the epoch change process
-        if current_committee.ready_to_change.len() > (2 * current_committee.members.len() / 3) {
-            // Todo: Reward nodes, choose new committee, increment epoch.
-            self.calculate_reputation_scores();
-            self.distribute_rewards();
-            // Todo: We can't really fail after here
-            // because changes have already been submitted above
-            // in the call to calculate_reputation_scores.
-            // Should we refactor change_epoch so it operates in two steps?
-            //  1. Validate all mutations that will be made and stage them.
-            //  2. Submit staged changes.
-            // Then, `clear_content_registry` could become
-            // `stage_clear_content_registry' and return the new state for the
-            // tables instead of applying the changes itself.
-            self.clean_up_content_registry();
-
-            // Clear executed digests.
-            self.executed_digests.clear();
-
-            self.committee_info.set(current_epoch, current_committee);
-            // Get new committee
-            let new_committee = self.choose_new_committee();
-            // increment epoch
-            current_epoch += 1;
-
-            // Set the new committee, epoch, and reset sub dag index
-            self.committee_info.set(current_epoch, new_committee);
-
-            self.metadata
-                .set(Metadata::Epoch, Value::Epoch(current_epoch));
-            TransactionResponse::Success(ExecutionData::EpochChange)
+        // If more than 2/3rds of the committee have signaled, start the committee selection
+        // process.
+        // Note that we do NOT want to execute this code after 2/3+1 nodes have signaled, since
+        // we only want to do it once to start the process.
+        if current_committee.ready_to_change.len() == (2 * current_committee.members.len() / 3) + 1
+        {
+            let start_block = self.get_block_number() + 1;
+            let end_block = start_block + self.get_commit_phase_duration();
+
+            tracing::info!(
```

Comment:
AFAIK this PR would be the first commits that we do logging inside an execution context. Also makes me a little nervous especially in cases like this one where we already set some state in the DB so we really need to return a Transaction::Success. Im not sure how safe writing to stdout is with tracing but we should make sure this is worth it and safe, because if there was any sort of panic here this node would be very out of sync. Curious on your thoughts about this.

Reply:
At this point we haven't committed anything to the DB, we've just staged it with atomo, and will get committed only when the atomo `run` finishes [here](https://github.com/fleek-network/lightning/blob/main/core/application/src/env.rs#L78). So in the event of a panic or even a node restart mid execution, it shouldn't result in partial changes being applied to the state in the DB. Does that make sense to you?

Reply:
Yeah i was forgetting about some of the safety atomo gives us. It does make sense to me